src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+            ISemanticFactsService semanticFacts,
+            CancellationToken cancellationToken,
+            out DocumentUpdateInfo documentUpdateInfo)
+                => new Converter(semanticModel, semanticFacts, cancellationToken).TryConvert(queryExpression, out documentUpdateInfo);
+
+        protected override QueryExpressionSyntax FindNodeToRefactor(SyntaxNode root, TextSpan span)
+        {
+            var node = root.FindNode(span);
+            return node as QueryExpressionSyntax ?? (node is ArgumentSyntax argument ? argument.Expression as QueryExpressionSyntax : default);
+        }
+
+        private sealed class Converter
+        {
+            private readonly SemanticModel _semanticModel;
+            private readonly ISemanticFactsService _semanticFacts;
+            private readonly CancellationToken _cancellationToken;
+
+            public Converter(SemanticModel semanticModel, ISemanticFactsService semanticFacts, CancellationToken cancellationToken)
+            {
+                _semanticModel = semanticModel;
+                _semanticFacts = semanticFacts;
+                _cancellationToken = cancellationToken;
+            }
+
+            public bool TryConvert(QueryExpressionSyntax source, out DocumentUpdateInfo documentUpdateInfo)
+            {
+                if (source.DescendantTrivia().Any(trivia => trivia.MatchesKind(
+                        SyntaxKind.SingleLineCommentTrivia,
+                        SyntaxKind.MultiLineCommentTrivia,
+                        SyntaxKind.MultiLineDocumentationCommentTrivia) ||
+                    source.ContainsDirectives))
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                // GetDiagnostics is expensive. Move it to the end if there were no bail outs from the algorithm.
+                // TODO likely adding more semantic checks will perform checks we expect from GetDiagnostics
+                // We may consider removing GetDiagnostics.
+                // https://github.com/dotnet/roslyn/issues/25639
+                if (!TryConvertInternal(source, out documentUpdateInfo) ||
+                    _semanticModel.GetDiagnostics(source.Span, _cancellationToken).Any())
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                return true;
+            }
+
+            private bool TryConvertInternal(QueryExpressionSyntax source, out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // (from a in b select a); 
+                SyntaxNode parent = source.WalkUpParentheses().Parent;
+
+                switch (parent.Kind())
+                {
+                    // return from a in b select a;
+                    case SyntaxKind.ReturnStatement:
+                        return TryConvertIfInReturnStatement(source, (ReturnStatementSyntax)parent, out documentUpdateInfo);
```

Comment:
>return TryConvertIfInReturnStatement(source, (ReturnStatementSyntax)parent, out documentUpdateInfo); [](start = 24, length = 100)

It looks like this is going to change the method to an iterator. Where do we check that the type of the method would allow that? Where do we check that this return statement is the only return statement in the method?

Reply:
@jinujoseph @DustinCampbell @dotnet/roslyn-ide 

I think we need to decide principles and write about it and make all refactorings consistent. if we decide to be strict on semantic, we should do what @AlekseyTs said. or if we decide to be permissive as we talked about, then we should allow even when there is existing return statement with warning/errors in preview window.